pkg/source/source.go

```diff
@@ -52,61 +58,68 @@ const (
 	// cilium network policy files from specific directory.
 	Directory Source = "directory"
+
+	// Please remember to add your source to defaultSources below.
 )
 
+// Sources is a priority-sorted slice of sources.
+type Sources []Source
+
+var defaultSources Sources = []Source{
+	KubeAPIServer,
+	Local,
+	KVStore,
+	CustomResource,
+	Kubernetes,
+	ClusterMesh,
+	LocalAPI,
+	Generated,
+	Restored,
+	Directory,
+	Unspec,
+}
+
 // AllowOverwrite returns true if new state from a particular source is allowed
 // to overwrite existing state from another source
 func AllowOverwrite(existing, new Source) bool {
-	switch existing {
-
-	// KubeAPIServer state can only be overwritten by other kube-apiserver
-	// state.
-	case KubeAPIServer:
-		return new == KubeAPIServer
-
-	// Local state can only be overwritten by other local state or
-	// kube-apiserver state.
-	case Local:
-		return new == Local || new == KubeAPIServer
-
-	// KVStore can be overwritten by other kvstore, local state, or
-	// kube-apiserver state.
-	case KVStore:
-		return new == KVStore || new == Local || new == KubeAPIServer
-
-	// Custom-resource state can be overwritten by other CRD, kvstore,
-	// local or kube-apiserver state.
-	case CustomResource:
-		return new == CustomResource || new == KVStore || new == Local || new == KubeAPIServer
-
-	// Kubernetes state can be overwritten by everything except clustermesh,
-	// local API, generated, restored and unspecified state.
-	case Kubernetes:
-		return new != ClusterMesh && new != LocalAPI && new != Generated && new != Restored && new != Unspec
-
-	// ClusterMesh state can be overwritten by everything except local API,
-	// generated, restored and unspecified state.
-	case ClusterMesh:
-		return new != LocalAPI && new != Generated && new != Restored && new != Unspec
-
-	// Local API state can be overwritten by everything except restored,
-	// generated and unspecified state
-	case LocalAPI:
-		return new != Generated && new != Restored && new != Unspec
-
-	// Generated can be overwritten by everything except by Restored and
-	// Unspecified
-	case Generated:
-		return new != Restored && new != Unspec
-
-	// Restored can be overwritten by everything except by Unspecified
-	case Restored:
-		return new != Unspec
-
-	// Unspecified state can be overwritten by everything
-	case Unspec:
-		return true
+	overflowNegative := overflowNegativeTo(len(sources))
+	return overflowNegative(slices.Index(sources, new)) <= overflowNegative(slices.Index(sources, existing))
+}
+
+func overflowNegativeTo(infinity int) func(int) int {
+	return func(n int) int {
+		if n < 0 {
+			return infinity
+		} else {
+			return n
+		}
 	}
+}
 
-	return true
+var Cell = cell.Module(
+	"source",
+	"Definitions and priorities of data sources",
+	cell.Provide(NewSources),
+)
+
+var sources Sources = defaultSources
+
+// MergeSourcesFunc enables modification of defaultSources.
+type MergeSourcesFunc func(existing Sources) Sources
```

Comment:
Where is this used?

Reply:
I don't see any meaningful usage in-tree:

```
$ git grep MergeSourcesFunc
pkg/source/source.go:// MergeSourcesFunc enables modification of defaultSources.
pkg/source/source.go:type MergeSourcesFunc func(existing Sources) Sources
pkg/source/source.go:   MergeFunc MergeSourcesFunc `optional:"true"`
pkg/source/source_test.go:      extraSrc := func() MergeSourcesFunc {
```

Reply:
There's no, but Jussi requested it in review above to be future-proof.

Reply:
If we can't articulate exactly why an extensible interface is needed, then I'm not sure we need an extensible interface. `Source` is about determining a priority order between conflicting resources that Cilium learns from different places. To be able to best reason about this order, it helps to explicitly outline what that source is, and in the commit message argue why that particular source is higher/lower priority that the other sources.

This effort is for XDS, yes? Would it satisfy the concerns if we added an XDS source into the priority list and assume that it has a particular priority compared with the other sources of information in LB?

Reply:
> If we can't articulate exactly why an extensible interface is needed, then I'm not sure we need an extensible interface.

The original argumentation was here: https://github.com/cilium/cilium/pull/35810#discussion_r1836161757.

> This effort is for XDS, yes? 

Roughly speaking, yes.

> Would it satisfy the concerns if we added an XDS source into the priority list and assume that it has a particular priority compared with the other sources of information in LB?

As for me, I'm completely fine with hardcoding these things at the appropriate time (or place). Privately, also @ovidiutirla expressed doubts about `MergeSourcesFunc` because not all callers of `AllowOverwrite()` are migrated to hive yet. So it seems it is 3:1 against @joamaki. Unless Jussi opposes, I'm happy to remove `MergeSourcesFunc` and the surrounding hive machinery.

Reply:
Done.